### PR TITLE
More tweaks towards support shared memory in parlang backends

### PR DESCRIPTION
**Pull request by FabioLuporini**
"parlang" is the new designated word to generically refer to low level languages for device programming, such as CUDA, HIP, SYCL. 

Note that currently for pragma-based programming models (OpenMP, OpenACC) we use "parpragma" 

Anyway, this PR is just yet another batch of fixes, tweaks, generalizations... nothing of particular relevance

### CODE REVIEW

File: tests/test_dle.py [hunk 1]
```diff
@@ -262,33 +262,61 @@ def test_cache_blocking_structure_optrelax_linalg(opt, expected):
...
+
+        bns, _ = assert_blocking(op, {'x0_blk0', 'x1_blk0'})
+        assert len(bns) == len(expected)
+        for root, v in zip(bns.values(), expected):
+            iters = FindNodes(Iteration).visit(root)
+            iters = [i for i in iters if i.dim.is_Block and i.dim._depth == 1]
+            assert len(iters) == len(v)
+            assert all(i.step == j for i, j in zip(iters, v))
+
+    def test_structure_2p5D(self):
```
Comment (georgebisbas) [inline tests/test_dle.py:295]: 2p5D?

File: tests/test_dle.py [hunk 2]
```diff
@@ -262,33 +262,61 @@ def test_cache_blocking_structure_optrelax_linalg(opt, expected):
...
+        par_tile = (16, 4)
 
-    eqns = [Eq(u.forward, u.dx),
-            Eq(v.forward, u.forward.dx)]
+        op = Operator(eqns, opt=('advanced', {'par-tile': par_tile,
+                                              'blockinner': True}))
 
-    op = Operator(eqns, opt=('advanced', {'par-tile': par_tile,
-                                          'blockinner': True}))
+        # 3D grid, but par-tile has only 2 entries => generates so called
```
Comment (georgebisbas) [inline tests/test_dle.py:309]: 2.5D blocking is usually used for 2D+simd, which is what by default would be constructed even without par-tile. 
Maybe rephrase this comment?

**Review by georgebisbas — CHANGES_REQUESTED** [review]
Need change of PR name

File: devito/mpi/halo_scheme.py [hunk 3]
```diff
@@ -113,15 +117,27 @@ def union(self, halo_schemes):
             for k, v in i.fmapper.items():
                 hse = fmapper.setdefault(k, v)
 
-                # The `loc_indices` must match
                 if hse.loc_indices != v.loc_indices:
-                    raise ValueError("Cannot compute the union of one or more HaloScheme "
-                                     "when the `loc_indices` differ")
+                    # The `loc_dirs` must match otherwise it'd be a symptom there's
+                    # something horrinbly broken somewhere!
```
Comment (georgebisbas) [inline devito/mpi/halo_scheme.py:122]: typos

File: tests/test_dle.py (hunk 2 above)
Comment (FabioLuporini) [reply tests/test_dle.py:309]: no why? I think the comment makes sense as it stands

File: tests/test_dle.py (hunk 1 above)
Comment (FabioLuporini) [reply tests/test_dle.py:295]: "2 point 5" -- makes sense, I think, as it's within the `TestBlockingParTile` class (so it's like "2.5D given par-tile")

File: devito/mpi/halo_scheme.py (hunk 3 above)
Comment (FabioLuporini) [reply devito/mpi/halo_scheme.py:122]: yes, fixing (these and those from the old PR in profiling.py, as promised)

File: tests/test_dle.py (hunk 1 above)
Comment (georgebisbas) [reply tests/test_dle.py:295]: ah ok, I thought it was referring to 2 dimensional 5 point jacobi stencil!

File: tests/test_dimension.py [hunk 4]
```diff
@@ -45,6 +45,38 @@ def test_multi_buffer_long_time(self):
         for i in range(7):
             assert np.allclose(g.data[i], 14+i+1)
 
+    def test_modulo_dims_generation(self):
+        grid = Grid((3, 3))
+
+        f = TimeFunction(name="f", grid=grid, save=Buffer(2))
```
Comment (mloubout) [inline tests/test_dimension.py:51]: Can we add a test that

```
f = TimeFunction(name="f", grid=grid, save=Buffer(2), time_order=2)
f.dt2 == ...
```

It works correctly but would be good to have it as a safety because this `Buffer(2)` will be mostly used in those cases

**Review by mloubout — APPROVED** [review]

**Review by georgebisbas — APPROVED** [review]

**Review by mloubout — APPROVED** [review]

**Comment by FabioLuporini** [comment]
Thanks everybody for the patience and the reviews. Now it's finally time to merge. Landing 🛩️ !
